File: servant/src/CardsProject/Marketplace/Events.hs
{-# LANGUAGE DeriveGeneric #-}
module CardsProject.Marketplace.Events where

import Data.Aeson (ToJSON, FromJSON)
import Data.Text (Text)
import GHC.Generics (Generic)

data OrderPaid = OrderPaid
  { orderId :: Int
  , playerId :: Int
  , total :: Double
  , paymentMethod :: Text
  , paidAt :: Text
  } deriving (Show, Eq, Generic)

instance ToJSON OrderPaid
instance FromJSON OrderPaid

data OrderShipped = OrderShipped
  { orderId :: Int
  , trackingNumber :: Text
  , shippedAt :: Text
  } deriving (Show, Eq, Generic)

instance ToJSON OrderShipped
instance FromJSON OrderShipped

data OrderRefunded = OrderRefunded
  { orderId :: Int
  , refundedAt :: Text
  } deriving (Show, Eq, Generic)

instance ToJSON OrderRefunded
instance FromJSON OrderRefunded

data TransactionCompleted = TransactionCompleted
  { transactionId :: Int
  , buyerId :: Int
  , sellerId :: Int
  , finalPrice :: Double
  , completedAt :: Text
  } deriving (Show, Eq, Generic)

instance ToJSON TransactionCompleted
instance FromJSON TransactionCompleted

File: laravel/app/Events/Marketplace/Events.php
<?php

namespace App\Events\Marketplace;

use Carbon\Carbon;

readonly class OrderPaid
{
    public function __construct(
        public readonly int $order_id,
        public readonly int $player_id,
        public readonly string $total,
        public readonly string $payment_method,
        public readonly \Carbon\Carbon $paid_at,
    ) {}
}

readonly class OrderShipped
{
    public function __construct(
        public readonly int $order_id,
        public readonly string $tracking_number,
        public readonly \Carbon\Carbon $shipped_at,
    ) {}
}

readonly class OrderRefunded
{
    public function __construct(
        public readonly int $order_id,
        public readonly \Carbon\Carbon $refunded_at,
    ) {}
}

readonly class TransactionCompleted
{
    public function __construct(
        public readonly int $transaction_id,
        public readonly int $buyer_id,
        public readonly int $seller_id,
        public readonly string $final_price,
        public readonly \Carbon\Carbon $completed_at,
    ) {}
}
